File: elixir/dnd-character/lib/dnd_character.ex
defmodule DndCharacter do
  defstruct ~w[strength dexterity constitution intelligence wisdom charisma hitpoints]a

  def modifier(score) do
    :math.floor((score - 10) / 2)
    |> trunc
  end

  def ability do
    for(_ <- 1..4, do: :rand.uniform(6))
    |> Enum.sort()
    |> Enum.drop(1)
    |> Enum.sum()
  end

  def character do
    constitution = ability()

    %DndCharacter{
      strength: ability(),
      dexterity: ability(),
      constitution: constitution,
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability(),
      hitpoints: 10 + modifier(constitution)
    }
  end
end

File: ruby/dnd-character/dnd_character.rb
class DndCharacter
  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  def initialize
    @strength = create_ability
    @dexterity = create_ability
    @constitution = create_ability
    @intelligence = create_ability
    @wisdom = create_ability
    @charisma = create_ability
    @hitpoints = 10 + DndCharacter.modifier(constitution)
  end

  def self.modifier(score) = (score - 10) / 2

  private
  def create_ability = (1..4).map { rand(6) + 1 }.sort.drop(1).sum
end
